### PR TITLE
Fixes #20773: Add a new logger to use properly use report_id

### PR DESCRIPTION
**Pull request by amousset**
https://issues.rudder.io/issues/20773

### CODE REVIEW

**Comment by amousset** [comment]
PR updated with a new commit

**Comment by amousset** [comment]
PR updated with a new commit

**Comment by amousset** [comment]
PR updated with a new commit

**Comment by amousset** [comment]
PR updated with a new commit

**Comment by amousset** [comment]
PR updated with a new commit

**Comment by amousset** [comment]
PR updated with a new commit

**Comment by amousset** [comment]
PR updated with a new commit

**Comment by amousset** [comment]
PR updated with a new commit

**Comment by amousset** [comment]
PR updated with a new commit

**Comment by amousset** [comment]
PR updated with a new commit

**Comment by amousset** [comment]
PR updated with a new commit

**Comment by amousset** [comment]
PR updated with a new commit

**Comment by amousset** [comment]
PR updated with a new commit

File: tree/20_cfe_basics/common.cf [hunk 1]
```diff
@@ -328,3 +328,64 @@ bundle agent ncf_classes_copy(source_prefix, destination_prefix)
...
+
+bundle agent call_method(method) {
+  vars:
+    !pass1::
+      # store initial value
+      "should_report" string => "${report_data.should_report}";
+    pass3::
+      "method_id"  string => "${report_data.method_id}";
+      # store value with new method_id
+      "call_method_data.${method_id}_${method}" string => "${should_report}";
```
Comment (ncharles) [inline tree/20_cfe_basics/common.cf:343]: It should be clearly stated that method and method_id can only use characters valids in classes name

File: tree/20_cfe_basics/log_rudder.cf [hunk 2]
```diff
@@ -815,3 +845,70 @@ bundle agent rudder_set_force_component_none(mode)
...
+
+# You must call pop_method_id later to remove it.
+# The start value is the report_id itself
+bundle agent method_id_push(method_id)
+{
+  vars:
+    # this must be evaluated exactly only once
+      "canon_method_id"       string => canonify("${method_id}");
+    pass3::
+      "old_value"             string => "${report_data.method_id}";
```
[the remark this answers is not in the document]
Comment (amousset) [reply tree/20_cfe_basics/log_rudder.cf:862]: no, the current value is in `report_data.method_id`, and we use the stack for old value (as a simple variable is easier to access).
I used this as it was already done this way for dry-run stack and I didn't want to reimplement this... _thing_

**Comment by amousset** [comment]
PR updated with a new commit

**Review by ncharles — APPROVED** [review]

**Comment by amousset** [comment]
PR updated with a new commit

**Pull request merged**